aosp_diff/preliminary/bionic/0006-bionic-libm-avx2-optimization.patch

```diff
@@ -0,0 +1,1045 @@
++    RETURN_FUNC(sinh_func, sinh_generic);
++}
++
++typedef double tanh_func(double __x);
++DEFINE_IFUNC_FOR(tanh) {
++    __builtin_cpu_init();
++    if (__builtin_cpu_supports("avx2")) RETURN_FUNC(tanh_func, tanh_avx2);
++    RETURN_FUNC(tanh_func, tanh_generic);
++}
++
++typedef double acos_func(double __x);
++DEFINE_IFUNC_FOR(acos) {
++    __builtin_cpu_init();
++    if (__builtin_cpu_supports("avx2")) RETURN_FUNC(acos_func, acos_avx2);
++    RETURN_FUNC(acos_func, acos_generic);
++}
++
++typedef float cosf_func(float __x);
++DEFINE_IFUNC_FOR(cosf) {
++    __builtin_cpu_init();
++    if (__builtin_cpu_supports("avx2")) RETURN_FUNC(cosf_func, cosf_avx2);
++    RETURN_FUNC(cosf_func, cosf_generic);
++}
++
++typedef float sinf_func(float __x);
++DEFINE_IFUNC_FOR(sinf) {
++    __builtin_cpu_init();
++    if (__builtin_cpu_supports("avx2")) RETURN_FUNC(sinf_func, sinf_avx2);
++    RETURN_FUNC(sinf_func, sinf_generic);
++}
++
++typedef double log_func(double __x);
++DEFINE_IFUNC_FOR(log) {
++    __builtin_cpu_init();
++    if (__builtin_cpu_supports("avx2")) RETURN_FUNC(log_func, log_avx2);
++    RETURN_FUNC(log_func, log_generic);
++}
++
++typedef long double logl_func(long double __x);
++DEFINE_IFUNC_FOR(logl) {
++    __builtin_cpu_init();
++    if (__builtin_cpu_supports("avx2")) RETURN_FUNC(logl_func, logl_avx2);
++    RETURN_FUNC(logl_func, logl_generic);
++}
++
++typedef float logf_func(float __x);
++DEFINE_IFUNC_FOR(logf) {
++    __builtin_cpu_init();
++    if (__builtin_cpu_supports("avx2")) RETURN_FUNC(logf_func, logf_avx2);
++    RETURN_FUNC(logf_func, logf_generic);
++}
++
++typedef void sincosf_func(float __y, float *__sinp, float *__cosp);
++DEFINE_IFUNC_FOR(sincosf) {
++    __builtin_cpu_init();
++    if (__builtin_cpu_supports("avx2")) RETURN_FUNC(sincosf_func, sincosf_avx2);
++    RETURN_FUNC(sincosf_func, sincosf_generic);
++}
++
++}  // extern "C"
+diff --git a/libm/x86/e_acos.S b/libm/x86/e_acos.S
+index 04b178719..c22012195 100644
+--- a/libm/x86/e_acos.S
++++ b/libm/x86/e_acos.S
+@@ -92,7 +92,7 @@ static_func:
+ # -- End  static_func
+ 
+ # -- Begin  acos
+-ENTRY(acos)
++ENTRY(acos_generic)
+ # parameter 1: 8 + %ebp
+ ..B2.1:
+ ..B2.2:
+@@ -384,7 +384,7 @@ ENTRY(acos)
+         popl      %ebp
+         ret       
+ ..B2.3:
+-END(acos)
++END(acos_generic)
+ # -- End  acos
+ 
+ # Start file scope ASM
+diff --git a/libm/x86/e_sinh.S b/libm/x86/e_sinh.S
+index b9a29308d..bde9a92fd 100644
+--- a/libm/x86/e_sinh.S
++++ b/libm/x86/e_sinh.S
+@@ -93,7 +93,7 @@ static_func:
+ # -- End  static_func
+ 
+ # -- Begin  sinh
+-ENTRY(sinh)
++ENTRY(sinh_generic)
+ # parameter 1: 8 + %ebp
+ ..B2.1:
+ ..B2.2:
+@@ -320,7 +320,7 @@ ENTRY(sinh)
+         popl      %ebp
+         ret       
+ ..B2.3:
+-END(sinh)
++END(sinh_generic)
+ # -- End  sinh
+ 
+ # Start file scope ASM
+diff --git a/libm/x86/kabylake/acos_avx2.c b/libm/x86/kabylake/acos_avx2.c
+new file mode 100644
+index 000000000..280eeaf7d
+--- /dev/null
++++ b/libm/x86/kabylake/acos_avx2.c
```

Comment:
Do we really need new files? Just check if this is the way we do in bionic. We have other options of using #defines and have a single .c file